sys/kern/vm_object.c

```diff
@@ -55,43 +55,36 @@ void vm_object_add_page(vm_object_t *obj, off_t offset, vm_page_t *pg) {
   }
 }
 
-static void vm_object_remove_page_nolock(vm_object_t *obj, vm_page_t *page) {
-  page->offset = 0;
-  page->object = NULL;
-
-  TAILQ_REMOVE(&obj->list, page, obj.list);
-  vm_page_free(page);
-  obj->npages--;
-}
-
-void vm_object_remove_page(vm_object_t *obj, vm_page_t *page) {
-  SCOPED_MTX_LOCK(&obj->mtx);
-
-  vm_object_remove_page_nolock(obj, page);
-}
-
-void vm_object_remove_range(vm_object_t *object, off_t offset, size_t length) {
-  SCOPED_MTX_LOCK(&object->mtx);
+static void vmo_remove_pages_nolock(vm_object_t *obj, vm_offset_t offset,
+                                    size_t length) {
+  assert(mtx_owned(&obj->mtx));
+  assert(page_aligned_p(offset) && page_aligned_p(length));
 
   vm_page_t *pg, *next;
-  TAILQ_FOREACH_SAFE (pg, &object->list, obj.list, next) {
-    if (pg->offset >= (off_t)(offset + length))
+  TAILQ_FOREACH_SAFE (pg, &obj->list, obj.list, next) {
+    if (pg->offset >= offset + length)
       break;
-    if (pg->offset >= offset)
-      vm_object_remove_page_nolock(object, pg);
+    if (pg->offset < offset)
+      continue;
+
+    pg->offset = 0;
+    pg->object = NULL;
+    TAILQ_REMOVE(&obj->list, pg, obj.list);
+    vm_page_free(pg);
+    obj->npages--;
   }
 }
 
+void vm_object_remove_pages(vm_object_t *obj, vm_offset_t off, size_t len) {
+  SCOPED_MTX_LOCK(&obj->mtx);
+  vmo_remove_pages_nolock(obj, off, len);
+}
+
 void vm_object_free(vm_object_t *obj) {
   if (!refcnt_release(&obj->ref_counter))
     return;
 
-  WITH_MTX_LOCK (&obj->mtx) {
-    vm_page_t *pg, *next;
-    TAILQ_FOREACH_SAFE (pg, &obj->list, obj.list, next)
-      vm_object_remove_page_nolock(obj, pg);
-  }
-
+  vm_object_remove_pages(obj, 0, -PAGESIZE);
```

Comment:
`len` argument in `vm_object_remove_pages` is `size_t`. Here you pass negative value. It's hard to understand without any description.

Reply:
I could use `rounddown(~0, PAGESIZE)` or `~0 & ~(PAGESIZE-1)` – each one of them is somewhat ugly.

Maybe a macro `vm_object_remove_all_pages(obj)` that expands to `vm_object_remove_pages(obj, 0, -PAGESIZE)` would do ?

Reply:
Macro is good enough.

Reply:
Done.